File: lessons/01-nodes/cpp/my_first_node.cpp
#include "rclcpp/rclcpp.hpp"

class MyNode:public rclcpp::Node{ //Inherit from the Node class provided by ROS2
    public:
        MyNode():Node("my_first_cpp_node"),counter_(0) { //Constructor , initialize the Node with the name "my_first_cpp_node" and initialize conter to 0
            RCLCPP_INFO(this->get_logger(),"Hello ROS2");//Log a message to the console
            timer_=this->create_wall_timer(std::chrono::seconds(1),//Creates a timer that calls the function every one second
            std::bind(&MyNode::timer_callback,this));//Binds the timer callback function to the timer
        }
    private:
        int counter_;
        void timer_callback() {//Fuction that is called every time the timer is triggered
            counter_++;
            RCLCPP_INFO(this->get_logger(),"Timer called %d times",counter_);
        }
        rclcpp::TimerBase::SharedPtr timer_;//Shared pointer to the timer object
};

int main(int argc, char **argv) {
    rclcpp::init(argc,argv); //nitialize ROS2
    auto node = std::make_shared<MyNode>(); //auto means we're bascally begging the compiler to figure out the type for  us
    rclcpp::spin(node);
    rclcpp::shutdown();
    return 0;
}

File: lessons/01-nodes/python/my_first_node.py
#!/usr/bin/env python3
import rclpy
from rclpy.node import Node


class MyNode(Node):
    def __init__(self): #Constructor of the MyNode class
        super().__init__("my_first_node") #Initialize the Node with the name "my_first_node"
        self.get_logger().info("Hello ROS2") #Log a message to the console
        self.create_timer(1.0,self.timer_callback) #Creates a timer that calls the function every 1 second
        self.counter_=0

    def timer_callback(self):
        self.get_logger().info(f"Timer callback called {self.counter_} times")
        self.counter_+=1

def main(args=None):
    rclpy.init(args=args) #Initialize the ROS2 system
    node=MyNode() #Create an instance of the MyNode class
    rclpy.spin(node) #Keep the node alive until Ctrl+C is pressed
    rclpy.shutdown() #Properly shutdown the ROS2 system

if __name__=="__main__":
    main()
